core/src/main/java/org/infinispan/protostream/annotations/impl/ProtoMessageTypeMetadata.java

```diff
@@ -834,7 +834,7 @@ protected XClass getJavaTypeFromAnnotation(ProtoField annotation) {
     * type.
     */
    private Object getDefaultValue(XClass clazz, String fieldName, XClass fieldType, Type protobufType, String value, boolean isRepeated) {
-      if (protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2) {
+      if (protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2 || protoSchemaGenerator.allowNullFields()) {
```

Comment:
this seems to do the trick:
```suggestion
      if (protoSchemaGenerator.syntax() == ProtoSyntax.PROTO2 || (protoSchemaGenerator.allowNullFields() && !fieldType.isPrimitive())) {
```